sktime/forecasting/compose/_networkpipeline.py

```diff
@@ -0,0 +1,269 @@
+            if argument_value == "original_X":
+                returned_arguments_kwarg[argument_name] = self._X
+                continue
+            if argument_value == "original_y":
+                returned_arguments_kwarg[argument_name] = self._y
+                continue
+            if argument_value == "original_fh":
+                returned_arguments_kwarg[argument_name] = self._fh
+                continue
+
+            if type(argument_value) == list:
+                out = []
+                for list_val in argument_value:
+                    returned_step_value = self._check_steps_for_values(list_val)
+                    if returned_step_value is not None:
+                        out.append(returned_step_value)
+                returned_arguments_kwarg[argument_name] = out
+                continue
+            # go through all steps and look for returned values
+            returned_step_value = self._check_steps_for_values(argument_value)
+            if returned_step_value is not None:
+                returned_arguments_kwarg[argument_name] = returned_step_value
+
+        return returned_arguments_kwarg
+
+    def fit(self, y, X=None, fh=None):
+        self._set_y_X(y, X)
+        self._set_fh(fh)
+        for name, alg, arguments in self._iter():
+            processed_arguments = {}
+
+            if "fit" in arguments and arguments["fit"] is None:
+                # this step must be skipped
+                continue
+            elif "fit" in arguments:
+                processed_arguments = self._process_arguments(arguments["fit"])
+            else:
+                processed_arguments = self._process_arguments(arguments)
+
+            if "y" in processed_arguments:
+                processed_arguments["y"] = check_y(processed_arguments["y"])
+            if "Z" in processed_arguments:
+                processed_arguments["Z"] = check_y(processed_arguments["Z"])
+            if "X" in processed_arguments and processed_arguments["X"] is not None:
+                processed_arguments["X"] = check_X(processed_arguments["X"])
+            if "X" in processed_arguments and processed_arguments["X"] is None:
+                del processed_arguments["X"]
+            # Transformers are instances of BaseTransformer and BaseEstimator
+            # Estimators are only instances of BaseEstimator
+            if hasattr(alg, "fit_transform"):
+                out = alg.fit_transform(**processed_arguments)
+                self._step_results[name] = out
+                self._is_fitted = True
+            processed_arguments["fh"] = self._fh
+            # estimators have fit and predict methods
+            if hasattr(alg, "fit") and hasattr(alg, "predict"):
+                alg.fit(**processed_arguments)
+                self._is_fitted = True
+
+        return self
+
+    def _predict(self, fh=None, X=None, return_pred_int=False, alpha=DEFAULT_ALPHA):
+        self._set_fh(fh)
```

Comment:
Having this line here isn't necessary because it's already being taken care of in the `predict` method which then calls the internal `_predict` method: https://github.com/alan-turing-institute/sktime/blob/f71f373aa8f89497d654661976d745c747c1cb0e/sktime/forecasting/base/_sktime.py#L206

Perhaps that fixes the other error?

Reply:
I removed the line `self._set_fh(fh)` from the `_predict` method but the test about the cutoff not having .freq attribute still fails.

Reply:
I'm not sure why this information is lost? I suggest to step through it with the debugger. I know that the functionality around the forecasting horizon is a bit brittle given the required date/time conversions but haven't seen this error before.

Reply:
It also only happens in specific cases in `update_predict`. The test is defined here: https://github.com/alan-turing-institute/sktime/blob/8d0b94b4992083f3c1e388afaeaf9f644759e68f/sktime/forecasting/tests/test_all_forecasters.py#L289

Reply:
Please also update with latest changes on `main`.

Reply:
@mloning I think I found the problem but I am not sure that I have the right solution.

The dataset that is being used for the test has daily observations starting on 2000-01-01 and finishing on 2000-02-06. When the second prediction is made, a prediction for 2000-02-11 is produced, i.e. skipping the dates 2000-02-07 to 2000-02-10.

When calling the line `self._update_y_X(y, X)` at the next `update()`, `self._y` is set with an `index.freq=None` I presume because of the gap in the index.

My temporary fix was to do:

```Python
y_index_frequency = self._y.index.freq
self._update_y_X(y, X)
self._y = self._y.asfreq(y_index_frequency)
```

This sets the index frequency to the correct one and the test passes but `self._y` has some missing values.

Do you think this behaviour is correct, i.e. `predict()` skipping certain days? For the purposes of this test I am just using a NaiveForecaster.

Reply:
I see, what I don't understand yet is, why this fails in your implementation but not the other algorithms?

Reply:
@mloning to be honest I am not 100% sure either, my implementation is fairly simple at the moment, I think.

Maybe this is linked to the chain of inheritance and method calls which I could not fully follow. While I was debugging the code I remember seeing that `predict()` was called more than once for the same date, this might be somehow linked to the issue but this would need to be investigated.

From what I see the code now passes the tests and can be merged. I will experiment with the pipeline some more with some real datasets to see whether it produces meaningful forecasts. Feel free to review and let me know if you have comments at this stage.